### DISCUSSION TITLE
Migration of Unit test framework from Karma+Jasmine to Jest framework proposal document

### DISCUSSION THREAD

**Discussion by arun1205** [Sunbird-coKreat/Community#38] [Feature Requests]
**Introduction**

This wiki recommends the migration of the current unit testing framework from Karma+Jasmine to a framework which is faster, compatible with newer versions of Ubuntu and angular. 

**Background & Problem statement**

Currently, the Sourcing platform’s code uses Karma+Jasmine as a unit testing framework which fails whenever a newer version of Ubuntu image/angular is used.

**Proposed Solution 1:**

Karma+Jasmine needs to be updated to a newer version, so that with every consequent update of Ubuntu/Angular, the testing framework stays consistent with the newer versions of Ubuntu/Angular.

As a caveat, it needs to be noted that a lot of reverse engineering efforts needs to done while completing this implementation 

Secondly, The time taken to run these tests after update will not be reduced substantially, since running of parallel tests is not efficient with Karma+Jasmine

 

**Proposed Solution 2:**

- Jest as a testing framework complies to all the aspects (faster, compatible with newer versions of Ubuntu and angular) and also runs parallel tests which results in better performance and less time taken to execute. 
- 
- Easy to install with zero setup hiccups. In addition, Jest provides you with an integrated framework that is ready to use, aiming for an out-of-the-box experience.
- It is maintained and well documented, making it easy to learn, especially for developers new to the TDD (Test Driven Development) approach.
- Jest is a fast-performing test automation framework where isolated tests are executed by performing [parallel testing](https://www.lambdatest.com/blog/what-is-parallel-testing-and-why-to-adopt-it/) in their processes.
- Jest provides testing support for asynchronous code.
- Auto mocking is another out-of-the-box feature of Jest. Jest can easily mock any dependency imported in the module that needs to be tested in the future.
- Jest library provides pre-commit hooks that allow running only relevant tests or running tests that have changed after the last commit.
- The code mods module offers many ways to integrate your current projects to sync with the latest framework seamlessly.

Other than the ones listed above, here are some more liked aspects of Jest

[Source](https://2018.stateofjs.com/testing/jest/)
Cons of Jest

- Jest has proven to be a powerful tool to do almost anything. However, while Jest has several advantages, it also has a few disadvantages. For example, it is an expensive framework, and at times its debugging capabilities need improvement. Below are a few cons of Jest:
- Compared to Jasmine and other frameworks, not many libraries and toolings are supported by Jest.
- People who are not comfortable with the Jest framework have asserted that the learning curve is pretty hard.
- Leveraging auto mocking can make tests slow. This is because the more dependencies a module has, the more work needs to be done by Jest to mock it, which comes at the cost of performance.
- Snapshot testing with Jest is not so feasible for larger snapshot files containing thousands of lines.

Other than the ones listed above, here are some of the disliked aspects of Jest among developers

[Source](https://2018.stateofjs.com/testing/jest/)
 
**Confluence document link** : https://project-sunbird.atlassian.net/wiki/spaces/~6172572fa9897100700e621d/pages/3254943768/Migration+of+Unit+test+framework+from+Karma+Jasmine+to+Jest+framework+proposal
Conclusion:

<TODO>
@vrayulu

**Comment by HarishGangula** [comment]
@arun1205  
We should stick to karma and jasmine for unit testing since Angular recommends
Angular developers know these frameworks by default changing the test framework causes confusion when issues with angular and Jest come up since there is no official document for same and also Jest learning curve is high also

For improving test execution performance did we try [this](https://blog.aakashgoplani.in/simple-configuration-hacks-to-improve-unit-test-case-performance-using-jasmine-and-karma-in-angular)?

For running this test case we can use nodejs image from circleci docker list which doesn't have any issues which are used by other sunbird BBs

**Reply by arun1205** [reply]
@HarishGangula We are already using the headless chrome , but have not implemented the parallel processing .. Will check with the process as suggested in [this](https://blog.aakashgoplani.in/simple-configuration-hacks-to-improve-unit-test-case-performance-using-jasmine-and-karma-in-angular)
